### PR TITLE
[SPARK-25272][PYTHON][TEST] Add test to better indicate pyarrow is installed and related tests will run

### PR DESCRIPTION
**Pull request by BryanCutler**
## What changes were proposed in this pull request?
Currently, Arrow tests in python are skipped if pyarrow is not installed and there is clear Jenkins outputs that these tests were skipped. When it is installed, the tests are run but there will be no output. This adds a test that will be skipped and output a message in the Jenkins log that pyarrow and pandas required versions are installed. This way it is clear that Arrow tests will be run.  The output given when pyarrow and pandas are installed is:
`test_required_pyarrow_pandas_not_installed (pyspark.sql.tests.HaveArrowTests) ... skipped 'Required pyarrow and pandas were found, Arrow tests will run'

## How was this patch tested?

Added new test

### CODE REVIEW

**Comment by BryanCutler** [comment]
@HyukjinKwon wdyt? This way there will be a clear log message that we can point to and say the Arrow tests were run.

**Comment by BryanCutler** [comment]
Hmmm, I don't see the skipped test which is weird.. let me look into that

**Comment by BryanCutler** [comment]
It prints in my local tests for py3.5, lets try again

**Comment by BryanCutler** [comment]
retest this please

**Comment by BryanCutler** [comment]
Ok, that last build worked, not sure what happened with the first
```
Skipped tests in pyspark.sql.tests with python3.4:
    test_required_pyarrow_pandas_not_installed (pyspark.sql.tests.HaveArrowTests) ... skipped 'Required pyarrow and pandas were found, Arrow tests will run'
    test_create_dataframe_required_pandas_not_found (pyspark.sql.tests.SQLTests) ... skipped 'Required Pandas was found.'
    test_to_pandas_required_pandas_not_found (pyspark.sql.tests.SQLTests) ... skipped 'Required Pandas was found.'
    test_type_annotation (pyspark.sql.tests.ScalarPandasUDFTests) ... skipped 'Type hints are supported from Python 3.5.'
```

**Comment by BryanCutler** [comment]
Well, the purpose was so I could point to a specific line in the Jenkins test log and say the Arrow tests ran for python 3.  Right now, you have to say they ran because the output doesn't say they were skipped - which isn't all that assuring.  Will you test coverage report show clearly if the Arrow tests were ran or skipped for each python version? If so then we probably don't need this..

**Comment by HyukjinKwon** [comment]
Hmmmmm .. what I was thinking was, we assume they are installed by default, and skipped tests are logged.

**Comment by squito** [comment]
I'm kinda 'meh' / -0 on this change.  My issue wasn't so much not seeing something printed out, it was more (a) python output isn't integrated into jenkins test reports and (b) I'm still learning my way around the pyspark code base (and how to even be an effective python dev in general, for that matter).  I'd rather have something more general, rather than a one-off for this specific test. (Like I commented on jira, just updating the jenkins builds to archive the right xml test report file at least, or better still integrate into the test report.)

but I'll also leave this up to others, as I'm still not very knowledgeable here.

**Comment by BryanCutler** [comment]
>Hmmmmm .. what I was thinking was, we assume they are installed and ran by default, and skipped tests are logged.

Yeah but for people who don't know these tests or how the _have_xxx flags work, having an additional positive confirmation that the Arrow tests are run is nice.  We already are doing this for pandas [here](https://github.com/apache/spark/blob/master/python/pyspark/sql/tests.py#L3236).

>I'd rather have something more general, rather than a one-off for this specific test. (Like I commented on jira, just updating the jenkins builds to archive the right xml test report file at least, or better still integrate into the test report.)

I still agree on these points and this isn't instead of those. This would allow you to look in the test results for Python 3 (which runs the arrow tests) and be able to check for sure they ran with one line.

That being said, this message only appeared in 1 of the 3 builds here so far, which is pretty weird.

File: python/pyspark/sql/tests.py [hunk 1]
```diff
@@ -4090,6 +4090,25 @@ def __init__(self, **kwargs):
                 _make_type_verifier(data_type, nullable=False)(obj)
 
 
+class HaveArrowTests(unittest.TestCase):
+
+    @unittest.skipIf(_have_pandas and _have_pyarrow,
+                     "Required PyArrow and Pandas were found, Arrow tests will run")
+    def test_required_pyarrow_pandas_not_installed(self):
```
Comment (BryanCutler) [inline python/pyspark/sql/tests.py]: @HyukjinKwon any idea why this wouldn't have printed a skip message on the last 3 out of 5 tests?

**Comment by HyukjinKwon** [comment]
> Yeah but for people who don't know these tests or how the _have_xxx flags work, having an additional positive confirmation that the Arrow tests are run is nice. We already are doing this for pandas here.

hmm .. I thought the current information is enough to indicate which Arrow or Pandas we would use and test. For instance,

```
Skipped tests in pyspark.sql.tests with pypy:
    test_createDataFrame_column_name_encoding (pyspark.sql.tests.ArrowTests) ... skipped 'Pandas >= 0.19.2 must be installed; however, it was not found.'
    ...

Skipped tests in pyspark.sql.tests with python2.7:
    test_createDataFrame_column_name_encoding (pyspark.sql.tests.ArrowTests) ... skipped 'Pandas >= 0.19.2 must be installed; however, your version was 0.16.0.'
    ...
```

Yea, looks we are already doing this for Pandas .. but actually I would rather remove that test too.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by BryanCutler** [comment]
> I thought the current information is enough to indicate which Arrow or Pandas we would use and test
Well yeah, it is when they are skipped but my point was that having an additional positive confirmation that the tests were run would be nice.  Maybe that's just me though, so we don't have to merge this. 

I am still a bit concerned why my skip message was only printed in 1 of the tests here, so I'll run a few more to see if I can figure it out.

**Comment by holdenk** [comment]
Jenkins retest this please.

File: python/pyspark/sql/tests.py (hunk 1 above)
Comment (holdenk) [reply python/pyspark/sql/tests.py]: So looking at https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95514/consoleFull it's printing the other skip method ( "Required PyArrow and Pandas not found") so maybe only some of the Jenkins workers have the correct version of PyArrow and Pandas installed? We should probably check that?

**Comment by BryanCutler** [comment]
restest this please

**Comment by BryanCutler** [comment]
retest this please

**Comment by holdenk** [comment]
How about if we checked an env variable like EXPECT_ARROW (and set it when tests are running in Jenkins) and if it's set but we don't have the required versions installed fail? That way we know if our code is being tested in CI?

**Comment by BryanCutler** [comment]
Hey @holdenk , yeah we could do that but I'm ok with the current way which is to only print if the tests are skipped and assume they ran otherwise. I just wanted to make sure that the weird behavior I saw earlier wasn't happening anymore.  Looks good so far, but I was trying to hit one more worker to check, then I'll close this out.

**Comment by BryanCutler** [comment]
retest this please

**Comment by BryanCutler** [comment]
retest this please

**Comment by BryanCutler** [comment]
retest this please

**Pull request closed without merging**